patroni/ha.py

```diff
@@ -20,31 +20,28 @@
 from .postgresql.misc import postgres_version_to_int
 from .postgresql.postmaster import PostmasterProcess
 from .postgresql.rewind import Rewind
+from .tags import Tags
 from .utils import polling_loop, tzutc
 
 logger = logging.getLogger(__name__)
 
 
-class _MemberStatus(NamedTuple):
-    """Node status distilled from API response:
+class _MemberStatus(Tags, NamedTuple('_MemberStatus',
+                                     [('member', Member),
+                                      ('reachable', bool),
+                                      ('in_recovery', Optional[bool]),
+                                      ('wal_position', int),
+                                      ('data', Dict[str, Any])])):
+    """Node status distilled from API response.
```

Comment:
It seems we are refactoring not only tags, but also `_MemberStatus`.
I guess we should describe this in the PR so it's included in the commit, I mean, that we took advantage of the change to also change this class.

Also, it seems we got rid of `dcs_last_seen`. Was that intentional, or are we missing to add the corresponding property into the class?

Reply:
Refactoring of the `_MemberStatus` was unavoidable. `Tags` class dictates that child classes must implement `tags` property, but such an attribute already existed in the `_MemberStatus`. Renaming it to `_tags` wasn't possible (NamedTuple doesn't allow to have attributes that start with underscore) and therefore I just decided to keep the whole JSON response in the `data` attribute.
`tags`, `timeline`, and `watchdog_failed` attributes are returned from the JSON response as it is, and therefore I converted them to properties.
`dcs_last_seen` attribute is not used anywhere in the code. IIRC it was added to `_MemberStatus` just because it could be useful in the future. In two years it is still not used, therefore I decided to remove it. If we need it one day - adding a property would be just 4 lines of code.

`reachable`, `in_recovery`, and `wal_position` attributes are kept as it is because their values are calculated based on some logic. For example https://github.com/zalando/patroni/pull/2823/files#diff-b6a3a0ef64783a6d32feb908d74a3a009e9f25008b8ca5313c9c9747e3b74a63R54 will raise an exception is there is no `wal` or `xlog` in the JSON document and as a result the `fetch_node_status()` will return `_MemberStatus.unknown(member)`.